add support for external server for serving the static assets

tardyp opened the pull request:
Implementation for #6307 

There is still the need to setup CORS in the static server as system.js is fetching
the scripts via xhr.

Using scriptLoad: true will not work here (yet) as it needs to have much more metadata.

This also adds a better approach for cacheBusting (using the git commit id, instead of Date.now())

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=6311) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=6311) before we can accept your contribution.<br/>

torkelo commented:
cache busting is handled by the javascript file is minified and revisioned  (gets a hash in it's name)

tardyp committed:
add support for external server for serving the static assets

There is still the need to setup CORS in the static server as system.js is fetching
the scripts via xhr.

Using scriptLoad: true will not work here (yet) as it needs to have much more metadata.

tardyp commented:
@torkelo thanks, this wasn't clear for me, as I am building from source for 32bit 

make build will not generate the production JS, I had to dig into the grunt file to finally find the grunt release task.

I uploaded a new version without the cacheBusting fix.
It is now working on my home server with static files served by another ISP.

tardyp commented on public/app/system.conf.js:
```diff
@@ -1,6 +1,15 @@
 System.config({
   defaultJSExtenions: true,
-  baseURL: 'public',
+  baseURL: getBaseURL(),
```
I am not sure if this change is necessary.
Do the plugin system uses this? in this case this is necessary. If its only the config for the boot.js generation, then I will remove the change.
Please advise.

torkelo replied:
you cannot change baseURL in system.conf some things are loaded async (plugins and some sections)

tardyp replied:
Dont we want to load those things from the static server? In this case, we need to have system.conf point to the static server.

So for the build, we need it to point to local filesystem 'public', but for the production, we need it to point to the static server.

tardyp commented:
@torkelo is there anything I need to do for this PR to be merged?
Please let me know.

torkelo commented on docs/sources/installation/configuration.md:
```diff
@@ -114,6 +114,23 @@ callback URL to be correct).
+For optimization you can put those files on a host optimized for serving static
+files (like a CDN or S3 like server).
+
+> **Note**  You will need to setup CORS headers on your static server to make this work
```
Bit confused about line :) the CORS headers would be needed on the grafana-server http api, not the static server

tardyp replied:
Yes.
Because system.js loads its JS using XHR, then the static server must have cors configured to enable XHR from the grafana server URL

torkelo replied:
oh, yea your right

torkelo commented on pkg/setting/setting.go:
```diff
@@ -172,7 +173,7 @@ func init() {
 }
 
-func parseAppUrlAndSubUrl(section *ini.Section) (string, string) {
+func parseAppUrlAndSubUrl(section *ini.Section) (string, string, string) {
```
this seems a bit strange, why mingle the static url parsing with the static url?

tardyp replied:
It is for generating the default value, which is based on the public_url

torkelo replied:
but seems strange to have that in this function, or am I missing something?

tardyp committed:
Review comment fixes

- remove setup of static_url in system.conf.js
  after discussion, this is not straightforward to load the plugins from a static server
- remove references to CORS.
  As system.js will not load anything anymore in production mode, there is no need to setup cors.
- mention that the plugins are still served by the main server
- refactor the setting code to parse static_url apart from the AppUrl

tardyp commented:
I updated the code to address the review comments

tardyp committed:
Update setting.go

torkelo commented:
Have you tested to install external plugins, does it work?

torkelo commented:
Have you tested to install external plugins, does it work?

tardyp commented:
@torkelo I have not yet tested the installed plugin

torkelo commented:
created feature request for this issue, https://github.com/grafana/grafana/issues/6844 

Closing this for now to gauge interest.

Gauravshah commented:
interesting, would help speed up load times.
